samples/event-sourcing-esdb-simple/src/main/java/io/eventdriven/ecommerce/core/entities/EntityStore.java

```diff
@@ -0,0 +1,96 @@
+package io.eventdriven.ecommerce.core.entities;
+
+import com.eventstore.dbclient.AppendToStreamOptions;
+import com.eventstore.dbclient.EventStoreDBClient;
+import com.eventstore.dbclient.ExpectedRevision;
+import io.eventdriven.ecommerce.core.http.ETag;
+import io.eventdriven.ecommerce.core.serialization.EventSerializer;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class EntityStore<TEntity> {
+  private final EventStoreDBClient eventStore;
+  private final BiFunction<TEntity, Object, TEntity> when;
+  private final Function<UUID, String> mapToStreamId;
+  private final Supplier<TEntity> getDefault;
+
+  public EntityStore(
+    EventStoreDBClient eventStore,
+    BiFunction<TEntity, Object, TEntity> when,
+    Function<UUID, String> mapToStreamId,
+    Supplier<TEntity> getDefault
+  ) {
+
+    this.eventStore = eventStore;
+    this.when = when;
+    this.mapToStreamId = mapToStreamId;
+    this.getDefault = getDefault;
+  }
+
+  public TEntity Get(UUID id) throws ExecutionException, InterruptedException {
+    var streamId = mapToStreamId.apply(id);
+    var result = eventStore.readStream(streamId).get();
+
+    var events = result.getEvents().stream()
+      .map(e -> EventSerializer.Deserialize(e))
+      .toList();
+
+    var current = getDefault.get();
+    for (var event : events) {
+      current = when.apply(current, event);
+    }
```

Comment:
w sumie to tutaj reduce byłby najlepszy

Reply:
Reduce is for associative operations (the order of how you combine the elements doesn't matter). In this case it matters right?

Reply:
Yup, reduce only allows transformation of the set of objects to the single object type, I could probably try to play and do collect, but IMHO that would be much less readable than just Ye Olde For each.

Reply:
I also forgot that there is no reduce in Java (we have something like that `fold` in Kotlin). 

I've used reduce in my repo for rehydrating state and it's not so readable :/ 
https://github.com/nowakprojects/DDD.EventSourcing.Java.Flights/blob/main/src/main/java/pl/zycienakodach/esflights/modules/ordering/domain/Ordering.java

But here... something like that should be OK:
```java
    return events.stream()
      .collect(getDefault, when::apply, when::apply);
```
But, regarding the third parameter:
you probably cannot use parallel stream here, because you need to keep ordering. 
So.... something like that need to be done :D 
```java
 return events.stream()
      .collect(getDefault, when::apply, (__,___) -> { throw new IllegalStateException("Parallel stream cannot be used for rehydrating the state"); });
```